README.adoc

```diff
@@ -36,30 +38,238 @@ include::https://raw.githubusercontent.com/spring-guides/getting-started-macros/
+The message acknowledgement mode is set in the adapter to automatic, by default. The example shows
+the acknowledgement mode set to manual just to show how to manually acknowledge messages since, in
+the automatic mode, nothing needs to be done.
 
+Attached to an inbound channel is a service activator which is used to process incoming messages.
 
-`src/main/java/hello/Application.java`
+`src/main/java/hello/PubSubApplication.java`
 [source,java]
 ----
-include::complete/src/main/java/hello/Application.java[]
+include::/complete/src/main/java/hello/PubSubApplication.java[tag=messageReceiver]
 ----
 
+The `ServiceActivator` input channel name must match the method name of the inbound channel declared
+above. In this example, the message body is simply logged and the message is acknowledged. The
+acknowledgement is done through the `AckReplyConsumer` object, which is sent in the message headers.
+
+== Create an outbound channel adapter
+
+An outbound channel adapter listens to new messages in an application's Spring channel and publishes
+them to a Google Cloud Pub/Sub topic.
+
+Instantiating an outbound channel adapter requires a `PubSubTemplate`, which is Spring's abstraction
+to publish messages to Google Cloud Pub/Sub topics. The Spring Cloud GCP Pub/Sub Boot starter
+provides a pre-configured `PubSubTemplate`.
+
+`src/main/java/hello/PubSubApplication.java`
+[source,java]
+----
+include::/complete/src/main/java/hello/PubSubApplication.java[tag=messageSender]
+----
+
+After the outbound channel adapter is instantiated, the name of the Google Cloud Pub/Sub topic
+("testTopic") where messages will be published to must be set. That topic must already exist before
+the application is run. New topics can be created through the
+https://console.cloud.google.com/cloudpubsub/topicList[Google Cloud Console] or, programatically,
+with the `PubSubAdmin` class.
+
+To write messages to a channel, in order to then publish them to Google Cloud Pub/Sub, a
+`MessagingGateway` can used.
+
+`src/main/java/hello/PubSubApplication.java`
+[source,java]
+----
+include::/complete/src/main/java/hello/PubSubApplication.java[tag=messageGateway]
+----
+
+From this code, Spring auto-generates an object that can then be autowired into a private field in
+our application.
+
+`src/main/java/hello/WebAppController.java`
+[source,java]
+----
+include::/complete/src/main/java/hello/WebAppController.java[tag=autowireGateway]
+----
+
+== Add controller login
+
+Add logic to your controller that lets you write to a Spring channel:
+
+`src/main/java/hello/WebAppController.java`
+[source,java]
+----
+include::complete/src/main/java/hello/WebAppController.java[]
+----
+
+== Authentication
+
+If the GOOGLE_APPLICATION_CREDENTIALS environment variable and the
```

Comment:
Redundant info, lets just point users to https://cloud.google.com/docs/authentication/getting-started and can be in a separate authentication section, mentioned just once towards the top of the guide.